testbed/tests/assertions.py

```diff
@@ -28,3 +28,73 @@ def assert_color(actual, expected):
                 expected.b,
                 approx(expected.a, abs=(1 / 255)),
             )
+
+
+def reverse_alpha_blending_over(blended_color, back_color, original_alpha):
+    # This is the reverse of the "over" operation and has
+    # been derived from the "over" operation formula, see:
+    # https://en.wikipedia.org/wiki/Alpha_compositing#Description
+
+    if original_alpha == 0:
+        return back_color
+    else:
+        front_color = rgba(
+            round(
+                (
+                    (blended_color.r * blended_color.a)
+                    - (back_color.r * back_color.a * (1 - original_alpha))
+                )
+                / original_alpha
+            ),
+            round(
+                (
+                    (blended_color.g * blended_color.a)
+                    - (back_color.g * back_color.a * (1 - original_alpha))
+                )
+                / original_alpha
+            ),
+            round(
+                (
+                    (blended_color.b * blended_color.a)
+                    - (back_color.b * back_color.a * (1 - original_alpha))
+                )
+                / original_alpha
+            ),
+            original_alpha,
+        )
+        return front_color
+
+
+def assert_background_color(actual, expected):
+    # For platforms where alpha blending is manually implemented, the
+    # probe.background_color property returns a tuple consisting of:
+    #   - The widget's background color
+    #   - The widget's parent's background color
+    #   - The widget's original alpha value - Required for deblending
+    if isinstance(actual, tuple):
+        actual_widget_bg, actual_parent_bg, actual_widget_bg_alpha = actual
+        deblended_actual_widget_bg = reverse_alpha_blending_over(
+            actual_widget_bg, actual_parent_bg, actual_widget_bg_alpha
+        )
+        if isinstance(expected, tuple):
+            expected_widget_bg, expected_parent_bg, expected_widget_bg_alpha = expected
+            deblended_expected_widget_bg = reverse_alpha_blending_over(
+                expected_widget_bg, expected_parent_bg, expected_widget_bg_alpha
+            )
+            assert actual_widget_bg_alpha == expected_widget_bg_alpha
+
+            if actual_widget_bg_alpha != 1 and expected_widget_bg_alpha != 1:
+                assert_color(deblended_actual_widget_bg, deblended_expected_widget_bg)
+
+            else:
+                assert_color(actual_widget_bg, expected_widget_bg)
+        else:
+            if expected == TRANSPARENT or expected.a == 0:
```

Comment:
I get why TRANSPARENT is a special case here (because it's going to be a string); but why differentiate "a=0"? Why not convert TRANSPARENT into "rgba(0,0,0,0)" and then use a single based on deblending? For that matter - shouldn't the math for a deblended color with alpha=1 result in the original color?

Reply:
>...but why differentiate "a=0"?...

Because we cannot deblend a color having an alpha value of 0 to get the exact original color:
https://github.com/beeware/toga/blob/3054df1e0579a871d4bed3bdba44dfad4ebbeeb6/testbed/tests/assertions.py#L41-L47
As, we can see in the above snippet, `original_alpha` with a value of 0 - will lead to a division by zero error. For preventing this, we are returning the `back_color`(parent color) in such cases:
https://github.com/beeware/toga/blob/3054df1e0579a871d4bed3bdba44dfad4ebbeeb6/testbed/tests/assertions.py#L38-L39
This works for asserting cases where the blended color returned by the probe can be deblended to the original color (i.e., the cases where the original alpha has a value in the range (0,1] ) - since the widget color would be the same as the parent color on alpha blending with alpha value of 0. 

But there is another set of tests that we are doing, where the expected value is not a tuple returned by the probe, but rather it is a predetermined test color data. 
https://github.com/beeware/toga/blob/3054df1e0579a871d4bed3bdba44dfad4ebbeeb6/testbed/tests/widgets/properties.py#L379-L384
In the above test, when there will be the case of color being `rgba(50, 128, 200, 0.0)` then we cannot use the following branch for assertion:
https://github.com/beeware/toga/blob/3054df1e0579a871d4bed3bdba44dfad4ebbeeb6/testbed/tests/assertions.py#L94-L95
because here `deblended_actual_widget_bg` would have the parent's background color instead of the predetermined expected color value.

Hence, we need to use this branch for such cases:
https://github.com/beeware/toga/blob/3054df1e0579a871d4bed3bdba44dfad4ebbeeb6/testbed/tests/assertions.py#L92-L93

Reply:
I added a comment explaining the reasoning behind the choice.